Fix mypy errors and improve the accuracy of the annotations

chadrik opened the pull request:
Changes:

- Settings have been added to `mypy.ini`
- Running `uvx mypy` passes without error
- Generic container classes now inherit from `typing.Generic`
- Added missing overloads
- Replaced use of `Generator` with `Iterator` (this is preferred unless using bidirectional generators)
- Lots of other accuracy changes that resulted in removing `type: ignore` comments

justinfx commented:
@chadrik have a few of these changes you introduced forced the minimum python requirement to 3.10? We still previously supported python 3.7 and have some older environments at my work that still use it. Are we expecting to drop older python support with this update?

chadrik commented:
I’m pretty sure I can change this so that it doesn’t affect the Python version requirements.  I’ll take a look soon.

justinfx commented:
@chadrik the mypy check currently fails because of the new strict option in `mypy.ini`:
https://github.com/justinfx/fileseq/actions/runs/17882659422/job/50863790115?pr=141#step:3:1227

chadrik commented on mypy.ini:
```diff
@@ -0,0 +1,5 @@
+[mypy]
+no_implicit_reexport=true
+strict=true
+files=src/fileseq
+python_version=3.9
```
done

chadrik commented:
I think the problem is that the mypy github action defaults to passing the root of the repo to mypy which is causing it to ignore the `files` section in mypy.ini, which in turn necessitated the `exclude` pattern.  I reverted the change where I removed the  `path` option for the mypy action, but I also added the exclude pattern just in case.
